proto-lens-protoc/src/Data/ProtoLens/Compiler/Definitions.hs

```diff
@@ -163,60 +177,67 @@ messageAndEnumDefs protoPrefix hsPrefix messages enums
 messageDefs :: Text -> String -> DescriptorProto
             -> [(Text, Definition Name)]
 messageDefs protoPrefix hsPrefix d
-    = thisDef : subDefs
+    = (protoName, thisDef)
+          : messageAndEnumDefs
+                (protoName <> ".")
+                hsPrefix'
+                (d ^. nestedType)
+                (d ^. enumType)
   where
-    protoName = d ^. name
+    protoName = protoPrefix <> d ^. name
+    hsPrefix' = hsPrefix ++ hsName (d ^. name) ++ "'"
     hsName n = unpack $ capitalize $ n
-    (ooFields, ooInfos) = mconcat [ (fs, [oo])
-                                  | (i, o) <- (zip [0..] (d ^. oneofDecl))
-                                  , let (fs, oo) = oneofInfo (o ^. name) i
-                                  ]
-    allFields = extractFields (\f -> isNothing (f ^. maybe'oneofIndex)) ++ ooFields
+    allFields = collectFieldsByOneofIndex (d ^. field)
+    thisDef =
+        Message MessageInfo
+            { messageName = fromString $ hsPrefix ++ hsName (d ^. name)
+            , messageDescriptor = d
+            , messageFields =
+                  map fieldInfo $ Map.findWithDefault [] Nothing allFields
+            , messageOneofFields =
+                  map (uncurry oneofInfo)
+                      $ zip [0..] $ map (^. name) $ d ^. oneofDecl
```

Comment:
D'oh, missed the obvious `zipWith`.  (Interestingly, hint actually recommended it, I just missed its message.)

Thanks for explaining how to use `^..`; I tried using `traverse` earlier but couldn't figure out the right invocation.